src/dynamicprompts/samplers/sampling_context.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import annotations
+
+import dataclasses
+from itertools import islice
+from random import Random
+from typing import TYPE_CHECKING, Iterable
+
+from dynamicprompts.commands import Command
+from dynamicprompts.enums import SamplingMethod
+from dynamicprompts.parser.config import ParserConfig, default_parser_config
+from dynamicprompts.samplers.consts import DEFAULT_RANDOM
+from dynamicprompts.types import StringGen
+from dynamicprompts.utils import squash_whitespace
+from dynamicprompts.wildcardmanager import WildcardManager
+
+if TYPE_CHECKING:
+    from dynamicprompts.samplers import Sampler
+
+
+def _build_default_samplers():
+    from dynamicprompts.samplers.combinatorial import CombinatorialSampler
+    from dynamicprompts.samplers.cycle import CyclicalSampler
+    from dynamicprompts.samplers.random import RandomSampler
+    return {
+        SamplingMethod.COMBINATORIAL: CombinatorialSampler(),
+        SamplingMethod.CYCLICAL: CyclicalSampler(),
+        SamplingMethod.RANDOM: RandomSampler(),
+    }
+
+
+@dataclasses.dataclass(frozen=True)
+class SamplingContext:
+    default_sampling_method: SamplingMethod
+    wildcard_manager: WildcardManager
+    samplers: dict[SamplingMethod, Sampler] = dataclasses.field(default_factory=_build_default_samplers)
+    ignore_whitespace: bool = False
+    parser_config: ParserConfig = default_parser_config
+    rand: Random = DEFAULT_RANDOM
+
+    def with_sampling_method(self, sampling_method: SamplingMethod) -> SamplingContext:
+        return dataclasses.replace(self, default_sampling_method=sampling_method)
+
+    @property
+    def default_sampler(self) -> Sampler:
+        return self.samplers[self.default_sampling_method]
+
+    def get_sampler_and_context(self, command: Command) -> tuple[Sampler, SamplingContext]:
+        """
+        Get the correct sampler instance and a sub-context (if necessary) for the given command.
+        """
+        command_sampling_method = command.sampling_method
+        if command_sampling_method is not None:
+            sampler = self.samplers[command_sampling_method]
+            context = self.with_sampling_method(command_sampling_method)
```

Comment:
You can't have a combinatorial context inside a nonfinite context., i.e. `{~A|B|{!X|Y}}` should be coerced to `{~A|B|{~X|Y}}`